server/join.go

```diff
@@ -29,36 +30,118 @@ import (
 const defaultDialTimeout = 30 * time.Second
 
 // TODO: support HTTPS
-func (cfg *Config) genClientV3Config() clientv3.Config {
+func genClientV3Config(cfg *Config) clientv3.Config {
 	endpoints := strings.Split(cfg.Join, ",")
 	return clientv3.Config{
 		Endpoints:   endpoints,
 		DialTimeout: defaultDialTimeout,
 	}
 }
 
-// prepareJoinCluster send MemberAdd command to pd cluster,
-// returns pd initial cluster configuration.
-func (cfg *Config) prepareJoinCluster() (string, error) {
+func memberAdd(client *clientv3.Client, urls []string) (*clientv3.MemberAddResponse, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
 	defer cancel()
 
-	client, err := clientv3.New(cfg.genClientV3Config())
+	return client.MemberAdd(ctx, urls)
+}
+
+func memberList(client *clientv3.Client) (*clientv3.MemberListResponse, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
+	defer cancel()
+
+	return client.MemberList(ctx)
+}
+
+// prepareJoinCluster send MemberAdd command to pd cluster,
+// returns pd initial cluster configuration.
+//
+// TL;TR: The join functionality is safe. With data, join does nothing, w/o data
+//        and it is not a member of cluster, join does MemberAdd, otherwise
+//        return an error.
+//
+// Etcd automatically re-join cluster if there is a data dir, so first we check
+// if there is data dir or not. With data dir, it returns an empty string(etcd
+// will get correct configurations from data dir.)
+//
```

Comment:
+// Etcd automatically re-joins the cluster if there is a data directory. So first it checks
 +// if there is a data directory or not. If there is, it returns an empty string (etcd
 +// will get correct configurations from the data directory.)